File: src/util/handlers/music.ts
// # ----------------------------------Imports --------------------------------- #
import { ExtendedClient, SkipEmbedBuilder } from "../structures"
import config from "../../config.json"
import { Player, SearchResult, Track, TrackUtils, UnresolvedTrack } from "erela.js";
import { CommandInteraction } from "discord.js";
import { VoiceHandler } from "./voice";
import { QueueHandler } from "./queue";
import { PlaylistEmbedBuilder, TrackEmbedBuilder } from "../structures";

// # ----------------------------------Config ---------------------------------- #

export class MusicHandler {

    private client: ExtendedClient;
    private VH: VoiceHandler
    QH: QueueHandler

    constructor(client: ExtendedClient) {
        this.client = client;
        this.VH = new VoiceHandler(client)
        this.QH = new QueueHandler(client)
    }

    private async addTrack (
        interaction: CommandInteraction, 
        player: Player, 
        track: Track | undefined=undefined, 
        tracks: Track[] | undefined=undefined, 
        result: SearchResult | undefined=undefined) {

        var embed = {}

        if (track) {
            embed = new TrackEmbedBuilder(interaction, track, player).toJSON()

            player.queue.add(track)
            console.log("Track added to queue");

        } else if (tracks && result) {
            embed = new PlaylistEmbedBuilder(interaction, result, player).toJSON()

            tracks.forEach(track => {
                player.queue.add(track)
            })
            console.log("Playlist added to queue");
        } 

        if (player.get<boolean>("idle")) {
            player.set("idle", false)
            player.setTrackRepeat(false)
            player.setVolume(config.music.volumeDefault)
            player.stop(1)
            
        } else if (!player.playing) {
            player.setVolume(config.music.volumeDefault)
            await player.play()
        } 
        
        interaction.editReply({embeds: [embed]})
    }

    async play(interaction: CommandInteraction, query: string) {
        const player = await this.VH.ensureVoice(interaction)
        if (!player){
            return

        } else {
            console.log(`Attempting to play song... Query: ${query}`)
        }
        
        await interaction.deferReply()

        const result = await player.search(query)
        console.log(result)

        try {

            switch (result.loadType) {
                case "LOAD_FAILED":
                    interaction.editReply("Failed to load track, please use a URL or different search term");
                    break;

                case "NO_MATCHES":
                    interaction.editReply("404 Song not found! Try something else.");
                    break;

                case "SEARCH_RESULT":
                    var track = result.tracks.at(0)
                    this.addTrack(interaction, player, track)
                    break;

                case "TRACK_LOADED":
                    var track = result.tracks.at(0)
                    this.addTrack(interaction, player, track)
                    break;

                case "PLAYLIST_LOADED":
                    var tracks = result.tracks;
                    this.addTrack(interaction, player, undefined, tracks, result)
                    break;

                default:
                    await interaction.editReply("Something unexpected happen. Contact your server owner immediately and let them know the exact command you tried to run.");
                    console.log(`Load type for play request defaulted. query '${query}' result as follows:`);
                    console.log(result);
                    break;
            }

        } catch (e) {
            console.log("Failed to play track. Error:" + e)
            return
        }

        this.QH.updatePages(player)

        return
    }
    async skip(interaction: CommandInteraction, index: number, trim_queue = true ) {
        if (!this.client.playerExists){
            await interaction.reply(":warning: There is no player running. Play a song first.")
            return
        }

        const player = await this.VH.ensureVoice(interaction)
        if (!player){
            return
        }

        if (player.queue.length == 0 && !player.trackRepeat) {
            await interaction.reply(":notepad_spiral: End of queue - time for your daily dose of idle tunes.")
            player.stop()
            return
        }

        if (player.trackRepeat) {
            const nextTrack = player.queue.current

            if (!nextTrack) {
                await interaction.reply(config.error.trackNotFound)
                return
            }

            const embed = new SkipEmbedBuilder(interaction, nextTrack, player, 0).toJSON()
            await interaction.reply({content: ":repeat_one: Repeat enabled - looping song.", embeds:[embed] }) 
            player.seek(0)
            console.log("Skipped (Repeating song).");
            return
        }
        
        if (index < 0) {
            await interaction.reply({content: ":warning: That index is too low! Queue starts at #1.",ephemeral: true})
            console.log(`Skip failed. Index too low(Expected: >=1. Recieved: ${index})`)
            return

        } else if (index > player.queue.length) {
            await interaction.reply({content: `:warning: That index is too high! Queue only ${player.queue.length} items long.`,ephemeral: true})
            console.log(`Skip failed. Index too low(Expected: <=${player.queue.length}. Recieved: ${index})`)
            return
        
        } else {

            if (trim_queue) {
                console.log(`Skipped queue to track ${index} of ${player.queue.length}`)

                if (index - 1 != 0) {
                    player.queue.remove(0, index - 1)
                }
            
            } else {
                console.log(`Jumped to track ${index} of ${player.queue.length} in queue.`)
                const jumpTrack = player.queue.remove(index - 1).at(0)

                if (!jumpTrack) {
                    await interaction.reply(config.error.trackNotFound)
                    return
                }

                player.queue.add(jumpTrack, 0) 
            }

            // resolve the track if it hasnt been done already
            var nextTrack = player.queue.at(0)
            if (TrackUtils.isUnresolvedTrack(nextTrack)) {
                nextTrack = await TrackUtils.getClosestTrack(nextTrack as UnresolvedTrack)
            }

            if (!nextTrack) {
                await interaction.reply(config.error.trackNotFound)
                return
            }

            player.stop()
            console.log("Skipped current track")

            const embed = new SkipEmbedBuilder(interaction, nextTrack, player, index).toJSON()
            await interaction.reply({embeds: [embed]}) 
            
            this.QH.updatePages(player) 
            await VoiceHandler.updateStatus(this.client, player)

            return
        }
    }
}

File: bot/handlers/music.py
import traceback

import discord
import lavalink

import config as cfg

from ..handlers.embeds import PlaylistEmbedBuilder, SkipEmbedBuilder, TrackEmbedBuilder
from ..handlers.presence import PresenceHandler
from ..handlers.queue import QueueHandler
from ..handlers.voice import VoiceHandler
from ..util import models


class MusicHandler:
    def __init__(self, bot: models.LavaBot) -> None:
        self.bot = bot
        self.voice_handler = VoiceHandler(self.bot)
        self.queue_handler = QueueHandler(self.bot, self.voice_handler)

    async def __add_track(
        self,
        interaction: discord.Interaction,
        player: lavalink.DefaultPlayer,
        track: lavalink.AudioTrack = None,
        tracks: list[lavalink.AudioTrack] = None,
        result: lavalink.LoadResult = None,
    ):
        embed = discord.Embed()

        if track:
            embed = TrackEmbedBuilder(interaction, track, player).construct()
            player.add(track)

        elif tracks and result:
            embed = PlaylistEmbedBuilder(interaction, result, player).construct()

            for track in tracks:
                player.add(track)

        if player.fetch("idle"):
            player.store("idle", False)
            player.set_loop(player.LOOP_NONE)
            await player.set_volume(cfg.player.volume_default)
            await player.skip()

        elif not player.is_playing:
            await player.set_volume(cfg.player.volume_default)
            await player.play()

        await interaction.followup.send(embed=embed)

    async def play(self, inter: discord.Interaction, search: str):
        player: lavalink.DefaultPlayer = await self.voice_handler.ensure_voice(inter)

        if "https://" not in search:
            search = "ytsearch:" + search

        player.node: lavalink.Node = player.node
        result: lavalink.LoadResult = await player.node.get_tracks(f"{search}")

        try:
            match result.load_type:
                case lavalink.LoadType.LOAD_FAILED:
                    await inter.followup.send(
                        content="Failed to load track, please use a different URL or different search term."
                    )

                case lavalink.LoadType.NO_MATCHES:
                    await inter.followup.send(
                        content="404 song not found! Try something else."
                    )

                case lavalink.LoadType.SEARCH | lavalink.LoadType.TRACK:
                    track = result.tracks[0]
                    await self.__add_track(inter, player, track)

                case lavalink.LoadType.PLAYLIST:
                    tracks = result.tracks
                    await self.__add_track(inter, player, None, tracks, result)

                case _:
                    await inter.followup.send(
                        content="Something unexpected happened. Contact your server owner or local bot dev(s) immediately and let them know the exact command you tried to run."
                    )

        except Exception as e:
            stacktrace = traceback.extract_stack(e.__traceback__.tb_frame)
            return

        self.queue_handler.update_pages(player)

        return

    async def skip(
        self, inter: discord.Interaction, index: int, trim_queue: bool = True
    ):
        player = await self.voice_handler.ensure_voice(inter)

        if len(player.queue) == 0 and player.fetch("idle"):
            await inter.followup.send(
                ":notepad_spiral: End of queue - time for your daily dose of idle tunes!"
            )
            await player.skip()
            return

        if player.loop == player.LOOP_SINGLE:
            next_track = player.current

            if not next_track:
                await inter.followup.send(
                    f"Error! Track not found. Somethign went wrong with playback - try kicking {cfg.bot.name} from the VC and trying again."
                )
                return

            embed = SkipEmbedBuilder(inter, next_track, player, 0)
            await inter.followup.send(
                ":repeat_one: Repeat enabled - looping song.",
                embed=embed.construct(),
            )
            await player.seek(0)
            return

        if index < 0:
            await inter.followup.send(
                ":warning: That index is too low! Queue starts at #1.", ephemeral=True
            )
            return

        elif index > len(player.queue):
            await inter.followup.send(
                f":warning: That index is too high! Queue only {len(player.queue)} items long.",
                ephemeral=True,
            )
            return

        else:
            if trim_queue:

                if index - 1 != 0:
                    player.queue = player.queue[index - 1 :]

            else:
                jump_track = player.queue.pop(index - 1)

                if not jump_track:
                    await inter.followup.send("Error! Track not found.")
                    return

                player.add(jump_track, index=0)

            next_track = player.queue[0]
            if isinstance(next_track, lavalink.DeferredAudioTrack):
                next_track = await next_track.load(self.bot.lavalink)
                try:
                    next_track = lavalink.decode_track(next_track)

                except Exception as e:
                    pass

            if not next_track:
                await inter.followup.send(":warning: Error! Track not found.")
                return

            await player.skip()

            embed = SkipEmbedBuilder(inter, player.current, player)
            await inter.followup.send(embed=embed.construct())

            self.queue_handler.update_pages(player)

            return
